worker/raft/raftbackstop/worker.go

```diff
@@ -0,0 +1,257 @@
+	return w.catacomb.Wait()
+}
+
+func (w *Worker) loop() error {
+	for {
+		select {
+		case <-w.catacomb.Dying():
+			return w.catacomb.ErrDying()
+		case details := <-w.serverDetails:
+			err := w.maybeRecoverCluster(details)
+			if err != nil {
+				return errors.Trace(err)
+			}
+		}
+	}
+}
+
+func (w *Worker) maybeRecoverCluster(details apiserver.Details) error {
+	if w.configUpdated {
+		w.config.Logger.Infof("raft configuration already updated; waiting for raft worker restart")
+		return nil
+	}
+	if len(details.Servers) != 1 {
+		return nil
+	}
+	if _, found := details.Servers[w.config.Tag.Id()]; !found {
+		return nil
+	}
+	if w.config.Raft.State() == raft.Leader {
+		return nil
+	}
+
+	raftServers, err := w.getConfiguration()
+	if err != nil {
+		return errors.Annotate(err, "getting raft configuration")
+	}
+
+	numServers := len(raftServers)
+	localServer := raftServers[raft.ServerID(w.config.Tag.String())]
+	if localServer == nil {
+		return nil
+	}
+
+	localVote := localServer.Suffrage != raft.Nonvoter
+	if numServers == 1 && localVote {
+		// The server can vote and has quorum, so it can become leader.
+		return nil
+	}
+
+	err = w.recoverCluster(localServer)
+	return errors.Annotate(err, "recovering cluster")
+}
+
+func (w *Worker) recoverCluster(server *raft.Server) error {
+	// Make a configuration that can be written into the log, and append it.
+	newServer := *server
+	newServer.Suffrage = raft.Voter
+	configuration := raft.Configuration{
+		Servers: []raft.Server{newServer},
+	}
+	w.config.Logger.Infof("appending recovery configuration to log: %#v", configuration)
+	data, err := encodeConfiguration(configuration)
+
+	// Work out the last term and index.
+	lastIndex, err := w.config.LogStore.LastIndex()
+	if err != nil {
+		return errors.Annotate(err, "getting last log index")
+	}
+	var lastLog raft.Log
+	err = w.config.LogStore.GetLog(lastIndex, &lastLog)
+	if err != nil {
+		return errors.Annotate(err, "getting last log entry")
+	}
+
+	// Prepare log record to add.
+	record := raft.Log{
+		Index: lastIndex + 1,
+		Term:  lastLog.Term,
+		Type:  raft.LogConfiguration,
+		Data:  data,
+	}
+	if err := w.config.LogStore.StoreLog(&record); err != nil {
+		return errors.Annotate(err, "storing recovery configuration")
+	}
+	w.config.Logger.Infof("configuration appended")
+	w.configUpdated = true
+	return nil
+}
+
+func encodeConfiguration(config raft.Configuration) ([]byte, error) {
+	buf := bytes.NewBuffer(nil)
+	hd := codec.MsgpackHandle{}
+	enc := codec.NewEncoder(buf, &hd)
+	err := enc.Encode(config)
+	return buf.Bytes(), err
+}
+
+func (w *Worker) getConfiguration() (map[raft.ServerID]*raft.Server, error) {
+	future := w.config.Raft.GetConfiguration()
+	err := w.waitFuture(future)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	servers := make(map[raft.ServerID]*raft.Server)
+	config := future.Configuration()
+	for i := range config.Servers {
+		// Use a local var so we don't get all entries pointing to
```

Comment:
It that an issue here? We're not using a loop variable, just an index

Reply:
Yeah - before this it was:
```go
	for _, server := range config.Servers {
		servers[server.ID] = &server
	}
```
But that meant that I got a map where all of the values were the same object. Which was really hard to debug. I can remove the comment if you want?

Reply:
Yes please,  and I don't think we need the extra var

Reply:
Not sure what you mean by extra var? Would you rather it was this?
```
servers[config.Servers[i].ID] = &config.Servers[i]
```
Comment ditched though.